packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,423 @@
+import {
+  QueryChildrenRequest,
+  QueryChildrenResponse,
+  QueryDataRequest,
+  QueryDataResponse,
+} from '@agoric/cosmic-proto/agoric/vstorage/query.js';
+import { encodeHex } from '@agoric/internal/src/hex.js';
+import { isStreamCell } from '@agoric/internal/src/lib-chainStorage.js';
+import { decodeBase64 } from '@endo/base64';
+import { Fail, makeError, q as quote } from '@endo/errors';
+
+/**
+ * @import {MinimalNetworkConfig} from '@agoric/client-utils/src/network-config';
+ * @import {JsonSafe} from '@agoric/cosmic-proto';
+ * @import {AbciQueryResponse} from '@cosmjs/tendermint-rpc';
+ *
+ * @typedef {bigint} Height;
+ *
+ * @typedef {JsonRpcResponse<{response: AbciQueryResponse}>} AbciResponse;
+ *
+ * @typedef {JsonRpcResponse<{
+ *  sync_info: {
+ *    catching_up: boolean;
+ *    latest_block_height: string;
+ *  }
+ * }>} NodeStatusResponse;
+ *
+ *
+ * @typedef {{
+ *  blockHeight: string;
+ *  values: Array<string>;
+ * }} StreamCell
+ */
+
+/**
+ * @template {any} T
+ * @typedef {{
+ *  id: number;
+ *  jsonrpc: string;
+ *  result: JsonSafe<T>
+ * }} JsonRpcResponse;
+ */
+
+/**
+ * @template {any} T
+ * @typedef {{
+ *  latest(height?: Height): Promise<Update<T>>;
+ * }} Topic<T>
+ */
+
+/**
+ * @template {any} T
+ * @typedef {Topic<T> & {
+ *  iterate(minimum?: Height, maximum?: Height): AsyncIterable<Update<T>>;
+ *  reverseIterate(maximum?: Height, minimum?: Height): AsyncIterable<Update<T>>;
+ * }} StreamTopic<T>
+ */
+
+/**
+ * @template {any} T
+ * @typedef {{
+ *  blockHeight: Height;
+ *  value: T;
+ * }} Update<T>
+ */
+
+export const INVALID_HEIGHT_ERROR_MESSAGE = 'Invalid height';
+export const INVALID_RPC_ADDRESS_ERROR_MESSAGE =
+  'No valid RPC address found in config';
+
+const MINIMUM_HEIGHT = 1n;
+
+export const PATH_PREFIX = '/agoric.vstorage.Query';
+
+export const PATHS = /** @type {const} */ ({
+  CHILDREN: 'Children',
+  DATA: 'Data',
+});
+
+const codecs = /** @type {const} */ ({
+  [PATHS.CHILDREN]: {
+    request: QueryChildrenRequest,
+    response: QueryChildrenResponse,
+  },
+  [PATHS.DATA]: {
+    request: QueryDataRequest,
+    response: QueryDataResponse,
+  },
+});
+
+/**
+ * @param {ReturnType<typeof parseValue>} value
+ */
+const assertValueStructure = value => {
+  if (!isStreamCell(value))
+    throw Error(`Expected a stream cell, got ${quote(value)}`);
+};
+
+/**
+ * @param {ReturnType<parseValue>} cell
+ */
+const isDataString = cell => !!cell && typeof cell === 'string';
+
+/**
+ * @param {{
+ *  errorMessage: string;
+ *  height?: Height;
+ *  kind: PATHS[keyof PATHS];
+ *  path: string;
+ *  rpcAddress?: MinimalNetworkConfig['rpcAddrs'][0];
+ * }} options
+ */
+const makeVstorageError = ({ errorMessage, height, kind, path, rpcAddress }) =>
+  makeError(
+    `Cannot read '${kind}' of '${path}' ${
+      rpcAddress ? `from '${rpcAddress}'` : ''
+    } ${height ? `at height '${height}'` : ''} due to error: ${errorMessage}`,
+  );
```

Comment:
Composite suggestion:
* Rename `kind` to `query` (also applies to the rest of the file) and `errorMessage` to `message`
* Rename `options` to `details`
* Make use of `@endo/errors`
* Throw the error rather than returning it
```suggestion
/**
 * @param {{
 *  message: string;
 *  height?: Height;
 *  query: PATHS[keyof PATHS];
 *  path: string;
 *  rpcAddress?: string;
 * }} details
 * @returns {never}
 */
const throwVstorageError = ({ message, height, query, path, rpcAddress }) =>
  Fail`Cannot read ${q(query)} of path ${q(path)} at ${q({ rpcAddress, height })} due to error ${q(message)}`;
```